File: pact-python/pact/discovery.py
"""PACT discovery document creation."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from .constants import PACT_VERSION


def create_discovery(
    site: str,
    schemas: List[str],
    endpoints: Dict[str, Any],
    description: Optional[str] = None,
    platforms: Optional[Dict[str, Any]] = None,
    rate_limit: Optional[Dict[str, Any]] = None,
    auth: Optional[Dict[str, Any]] = None,
    license: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    """Create a PACT discovery document for ``/.well-known/pact.json``."""
    discovery: Dict[str, Any] = {
        "pact": PACT_VERSION,
        "site": site,
        "schemas": schemas,
        "endpoints": endpoints,
    }
    if description is not None:
        discovery["description"] = description
    if platforms is not None:
        discovery["platforms"] = platforms
    if rate_limit is not None:
        discovery["rate_limit"] = rate_limit
    if auth is not None:
        discovery["auth"] = auth
    if license is not None:
        discovery["license"] = license
    return discovery

File: packages/core/src/discovery.ts
import type { PactDiscovery, PactEndpointDef, PactPlatformDef } from './types.js';
import { PACT_VERSION } from './constants.js';

export interface CreateDiscoveryOptions {
  site: string;
  description?: string;
  schemas: string[];
  endpoints: Record<string, PactEndpointDef>;
  platforms?: Record<string, PactPlatformDef>;
  rateLimit?: { rpm: number; burst?: number };
  auth?: { type: string; register?: string };
  license?: { ai_input: boolean; ai_train: boolean; attribution: boolean };
}

/**
 * Create a PACT discovery document for /.well-known/pact.json.
 */
export function createDiscovery(options: CreateDiscoveryOptions): PactDiscovery {
  const discovery: PactDiscovery = {
    pact: PACT_VERSION,
    site: options.site,
    schemas: options.schemas,
    endpoints: options.endpoints,
  };

  if (options.description !== undefined) {
    discovery.description = options.description;
  }

  if (options.platforms !== undefined) {
    discovery.platforms = options.platforms;
  }

  if (options.rateLimit !== undefined) {
    discovery.rate_limit = options.rateLimit;
  }

  if (options.auth !== undefined) {
    discovery.auth = options.auth;
  }

  if (options.license !== undefined) {
    discovery.license = options.license;
  }

  return discovery;
}
